src/main/kotlin/org/mentionbattle/snadapter/impl/socialnetworks/handlers/TwitterHandler.kt

```diff
@@ -1,36 +1,100 @@
-import org.mentionbattle.snadapter.impl.eventsystem.ExitEvent
-import org.mentionbattle.snadapter.impl.eventsystem.PrimitiveEventQueue
-import org.mentionbattle.snadapter.impl.eventsystem.LogEvent
+import org.mentionbattle.snadapter.impl.socialnetworks.initalizers.Tags
+import org.mentionbattle.snadapter.impl.socialnetworks.initalizers.TwitterTokens
 
+import org.mentionbattle.snadapter.impl.eventsystem.*
+import java.util.*
+import twitter4j.Status;
+import twitter4j.conf.ConfigurationBuilder;
+import twitter4j.StallWarning
+import twitter4j.StatusDeletionNotice
+import twitter4j.StatusListener
+import twitter4j.FilterQuery
+import twitter4j.TwitterStreamFactory
+import twitter4j.TwitterStream
 
 @SocialNetwork("Twitter")
-internal class TwitterHandler(eventQueue : PrimitiveEventQueue) : SocialNetworkHandler, EventHandler {
+internal class TwitterHandler(token: TwitterTokens, tags: Tags, eventQueue: PrimitiveEventQueue) : SocialNetworkHandler, EventHandler {
 
     val eventQueue = eventQueue
     var isWorking = true
+    val tokens = token
+    val tags = tags
+    lateinit var twitterStream: TwitterStream
+
+    private val listener = object : StatusListener {
+        override fun onStatus(status: Status) {
+            val url = "https://twitter.com/" + status.user.screenName + "/status/" + status.id.toString()
+            var contenderIds = intArrayOf()
+            for (key in tags.contenderA) {
+                if (status.text.contains(key, true)) {
+                    contenderIds = contenderIds.plus(1)
+                    break
+                }
+            }
+            for (key in tags.contenderB) {
+                if (status.text.contains(key, true)) {
+                    contenderIds = contenderIds.plus(2)
+                    break
+                }
+            }
+            for (id in contenderIds) {
+                eventQueue.addEvent(MentionEvent(id, "twitter",
+                        url, status.user.name, status.text,
+                        status.user.profileImageURLHttps.toString(), status.createdAt))
+            }
+            println("${status.text} $url")
+        }
+
+        override fun onDeletionNotice(statusDeletionNotice: StatusDeletionNotice) {}
+
+        override fun onTrackLimitationNotice(numberOfLimitedStatuses: Int) {
+            println("Got track limitation notice:" + numberOfLimitedStatuses)
+        }
+
+        override fun onScrubGeo(userId: Long, upToStatusId: Long) {}
+
+        override fun onStallWarning(warning: StallWarning) {
+            println("Got stall warning:" + warning)
+        }
+
+        override fun onException(ex: Exception) {
+            ex.printStackTrace()
+        }
+    }
+
+
     override fun handleEvent(event: Event) {
         when (event) {
             is ExitEvent -> {
-                isWorking = false
+                twitterStream.cleanUp(); // shutdown internal stream consuming thread
```

Comment:
The resourced will not free, if we crash in processData or we will not shutdown twitterStream if  cleanUp crashes

Reply:
If application crashes, then all the resources we've got will be freed. Am I right? Or I missed something?

Reply:
@novikd 
We are launching each handler in seperated coroutines, so if coroutine crashed, the application will continue work and we will lost resources.
More than, I want to a possibility to restart a handler if it crashed, so it significant to free all resources after crash.